microprofile/metrics-smallrye/src/main/java/org/wildfly/extension/microprofile/metrics/MetricCollector.java

```diff
@@ -88,6 +94,26 @@ public MetricRegistration collectResourceMetrics(final Resource resource,
                                                      Function<PathAddress, PathAddress> resourceAddressResolver) {
         MetricRegistration registration = new MetricRegistration();
         collectResourceMetrics0(resource, managementResourceRegistration, EMPTY_ADDRESS, resourceAddressResolver, registration);
+        // Defer the actual registration until the server is running and they can be collected w/o errors
+        PropertyChangeListener listener = new PropertyChangeListener() {
+            @Override
+            public void propertyChange(PropertyChangeEvent evt) {
+                if (ControlledProcessState.State.RUNNING == evt.getNewValue()) {
+                    registration.register();
+                } else if (ControlledProcessState.State.STOPPING == evt.getNewValue()) {
+                    // Unregister so if this is a reload they won't still be around in a static cache in MetricsRegistry
+                    // and cause problems when the server is starting
+                    registration.unregister();
+                    processStateNotifier.removePropertyChangeListener(this);
+                }
+
+            }
+        };
+        this.processStateNotifier.addPropertyChangeListener(listener);
+        // If server is already running, we won't get a change event so register now
+        if (ControlledProcessState.State.RUNNING == this.processStateNotifier.getCurrentState()) {
+            registration.register();
```

Comment:
@bstansberry it's probably a stupid question, but is there a risk of registering the metric twice if the state flips to RUNNING after listener is added but before the condition?

Reply:
@spyrkob No, there isn't, but it's probably better to code defensively against it by making the method synchronized.

It's not a stupid question. :)

This method is called in the context of a management op that modifying the system. The switch to RUNNING also occurs in the context of such a management op (a big composite op that runs during boot) and the switch to RUNNING doesn't happen until the code that would call this method has completed. The system won't allow concurrent execution of 2 such management ops.

Reply:
Great, thanks for the explanation :)

Reply:
@spyrkob I added the synchronized